request.go

```diff
@@ -58,6 +60,41 @@ func NewRequest(method, path string) *Request {
 	return &Request{Method: method, Filepath: cleanPath(path)}
 }
 
+// Context returns the request's context. To change the context,
+// use WithContext.
+//
+// The returned context is always non-nil; it defaults to the
+// background context.
+//
+// For incoming server requests, the context is canceled when the
+// client's connection closes.
+func (r *Request) Context() context.Context {
+	if r.ctx != nil {
+		return r.ctx
+	}
+	return context.Background()
```

Comment:
In general, one should never be returning the background context.

If no context has been set, then this should return nil, and this should indicate that they should not attempt to use ctx.Done().

Imagine here a program failed to set a context, but then used `<-ctx.Done()` to wait for the context to end…

Reply:
Hmm, I was just copying from net/http/request.go:

https://golang.org/src/net/http/request.go?s=11390:11433#L306

I think in general if you have a nil context it's the same as having context.Background() since it will never be cancelled. I'm willing to change this but figured we were mimicking http.Request so might want to keep the same properties around?

Reply:
Ah, didn’t know about that code. Looks good in that case, but… eh… just hangups from my days back in Google where the guidelines basically said, “never use context.Background(), if you’re using context.Background, you’re doing something wrong. (some exceptions apply)”

But then, this is building things from the bottom up, where the exceptions almost certainly apply…